src/marqo/core/inference/models/hugging_face_model.py

```diff
@@ -0,0 +1,201 @@
+        except (OSError, ValueError, RuntimeError) as e:
+            raise InvalidModelPropertiesError(
+                f"Marqo encountered an error loading the Hugging Face model, modelProperties={self.model_properties}. "
+                f"Please ensure that the model is a valid Hugging Face model and retry.\n"
+                f" Original error message = {e}") from e
+        return model, tokenizer
+
+    def _load_from_zip_file(self) -> Tuple:
+        """Load the model from a zip file."""
+        zip_file_path = download_model(
+            repo_location=self.model_properties.model_location,
+            url=self.model_properties.url,
+            auth=self.model_auth,
+            download_dir=ModelCache.hf_cache_path
+        )
+
+        model_dir = extract_huggingface_archive(zip_file_path)
+        try:
+            model = AutoModel.from_pretrained(model_dir).to(self.device)
+            tokenizer = AutoTokenizer.from_pretrained(model_dir)
+        except (OSError, ValueError, RuntimeError) as e:
+            raise InvalidModelPropertiesError(
+                f"Marqo encountered an error loading the Hugging Face model, modelProperties={self.model_properties}. "
+                f"Please ensure that the provided zip file is valid.\n"
+                f" Original error message = {e}") from e
+        return model, tokenizer
+
+    def _load_pooling_method(self) -> Callable:
+        """Load the pooling method for the model."""
+        if self.model_properties.pooling_method == PoolingMethod.Mean:
+            return _average_pool_func
+        elif self.model_properties.pooling_method == PoolingMethod.CLS:
+            return _cls_pool_func
+        else:
+            raise ValueError(f"Invalid pooling method: {self.model_properties.pooling_method}")
+
+    def encode(self, sentence: Union[str, List[str]], normalize=True, **kwargs) -> Union[FloatTensor, np.ndarray]:
+        if isinstance(sentence, str):
+            sentence = [sentence]
+
+        if self.model is None:
+            self.load()
+
+        self.model.normalize = normalize
+        tokenized_texts = self.tokenizer(
+            sentence,
+            padding=True,
+            truncation=True,
+            max_length=self.model_properties.token,
+            return_tensors="pt"
+        ).to(self.device)
+
+        with torch.no_grad():
+            model_output = self.model(**tokenized_texts)
+
+        attention_mask = tokenized_texts['attention_mask']
+
+        embeddings = self.pooling_func(model_output, attention_mask)
+
+        if normalize:
+            embeddings = F.normalize(embeddings, p=2, dim=1)
+
+        return self._convert_output(embeddings)
+
+    def _convert_output(self, output):
+        if self.device == 'cpu':
+            return output.numpy()
+        elif self.device.startswith('cuda'):
+            return output.cpu().numpy()
+
+def extract_huggingface_archive(path: str) -> str:
```

Comment:
Is this internal to the class? Why is it not under the class?

Reply:
I also make it a static method because these methods don't need to access instance-specific data.